docs/labs/redos.html

```diff
@@ -0,0 +1,159 @@
+<p>
+In this exercise we are going to fix an input validation that uses regular expressions but is susceptible to ReDos attack.
+
+
+<p>
+<h2>Task Information</h2>
+<p>
+
+<p>
+The code below sets up handlers for a get request on path <tt>/parts</tt>. This code could be triggered, for example, by requesting <tt>http://localhost:3000/part?id=AB123</tt> (if it was running at localhost and responding to port 3000). If there are no validation errors, the code is supposed to show the part id. If there is a validation error, it responds with HTTP error code 422 ("Unprocessable Content"), a status code suggesting that the request was invalid for some reason, along with an error message. 
+<p>
+For this lab we want to fix a regex that is susceptible to a ReDos attack, applying these countermeasures: 
+<ol>
+<li>Limit the maximum length of input strings and check the input length first. 
+  <ol>
+    <li>After selecting the id parameter (<tt>query('id')</tt>), use the validation requirement <tt>isLength()</tt> to limit the string size. We'll need to use the maximum optional parameter to set the maximum size required: <tt>isLength({max: YOUR_MAXIMUM})</tt></li>
+  </ol>
+</li>
+<li>Modify the regex so that it doesn’t have the worst-case behavior. Be especially wary of any group “(...)” that contains a branch and/or ends with a repeat and is itself repeated</li>
+</ol>
+
+
+<p>
+<h2>Interactive Lab (<span id="grade"></span>)</h2>
+<p>
+Change the code below, adding the mitigation steps to avoid ReDos:
+<ol>
+  <li>Limit the maximum length of input strings so the query parameter id is only accepted if it's no longer than <b>50</b> characters</li>
+  <li>Modify the regex so that it doesn’t have the worst-case behavior. The regex should meet this application's part id format requirement that is: one or more uppercase alphanumeric characters.</li>
+</ol>
+<form id="lab">
+  <pre><code
+  >// Set up Express framework and express-validator library
+  const express = require("express");
+  const app = express();
+  const { query, matchedData, validationResult } =
+      require('express-validator');
+  
+  // Implement requests, e.g., http://localhost:3000/parts?id=1
+  app.get('/parts',
+  <input id="attempt0" type="text" size="65" spellcheck="false"
+   value="  query('id').matches( /^([A-Z0-9]+)+$/ ) ,">
+    (req, res) =&gt; { // Execute this code if /parts seen
+      const result = validationResult(req); // Retrieve errors
+      if (result.isEmpty()) { // No errors
+        const data = matchedData(req); // Retrieve matching data
+        return res.send(`You requested part id ${data.id}!`);
+      }
+      res.status(422).send(`Invalid input`);
+    })
+  </code></pre>
+  <button type="button" class="hintButton">Hint</button>
+  <button type="button" class="resetButton">Reset</button>
+  <button type="button" class="giveUpButton">Give up</button>
+  <br><br>
+  <p>
+  <i>This lab was developed by David A. Wheeler at
```

Comment:
This lab wasn't developed by David A. Wheeler nor the Linux Foundation.
Please give yourself & your organization credit, you've earned it!